java-core/google-cloud-core/src/main/java/com/google/cloud/Timestamp.java

```diff
@@ -54,7 +54,7 @@ public final class Timestamp implements Comparable<Timestamp>, Serializable {
       new DateTimeFormatterBuilder()
           .appendOptional(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
           .optionalStart()
-          .appendOffsetId()
+          .appendZoneOrOffsetId()
```

Comment:
qq, why did this change?

Reply:
This is the method I found that could fix the tests. Honestly I did not dig deeper. The javadoc of the new method include:

>The offset ID parsing is equivalent to using appendOffset(String, String) using the arguments 'HH:MM:ss' and the no offset string '0'

PS: I did retry the tests with the original `appendOffsetId()` method and now they work. You will see that [this commit](https://github.com/googleapis/sdk-platform-java/pull/3330/commits/34bb09b9b7f3d821ddc9b7472afbe6b477b7d985) used `appendOffsetId()` and ran into offset parsing problems in the tests (the next commit introduces `asspendZoneAndOffsetId()`. I could also reproduce that problem locally by then.

I now restored it to `appendOffsetId()`

Reply:
After restoring, the [checks](https://github.com/googleapis/sdk-platform-java/actions/runs/11673610643/job/32504691386?pr=3330) fail like last time:

```
Error:  Failures: 
Error:    TimestampTest.parseTimestampWithTimeZoneOffset:245 value of: parseTimestampDuration(...)
expected: 2020-07-10T21:03:00Z
but was : 2020-07-10T14:03:00Z
```

Reply:
I confirmed that this test passes in java 11. I could not find any release note between 8 and 11 talking about parsing behavior.

Reply:
Hmm, it seems worrying that switching from threeten to java.time would cause a difference in behavior only for Java 8. Let's do some digging into `appendOffsetId`. Maybe it's a known regression in the JDK, but just want to check that `appendZoneOrOffsetId()` wouldn't break anything.

Reply:
Ok, looks like it all boils down to https://github.com/adoptium/jdk/commit/c6d209b505932b43d58502b119b86a6fa5b5338e

That commit (in Java 9) was a fix meant for [this issue](https://bugs.openjdk.org/browse/JDK-8066982)

At least we now know that it's a bug. The affected line is `Timestamp.java:215`:
https://github.com/googleapis/sdk-platform-java/blob/685a3e91236f29c29cae42cc37ccdd1ed30a1e54/java-core/google-cloud-core/src/main/java/com/google/cloud/Timestamp.java#L213-L217

Next is to figure out why `appendZoneOrOffsetId` works:

What I know is that JDK 8 will [correctly use](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/Parsed.java#L589-L591) the `zone` property to compute the epoch seconds. However, `zone` is only set when using [ZoneIdPrinterParser](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L3776) ([this](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L3908) is where the zone is set based on the offset via [`setParsed(ZoneId)`](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeParseContext.java#L417))

`appendZoneOrOffsetId` [computes](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L1114) offset strings (e.g. `-01:00`) into `ZoneOffset` classes in its [parsing function](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L3907) (i.e. computes a `zone` from the offset string). I think that's why: offsets are parsed into zones when using this `PrinterFormatter`, whereas the [`OffsetIdPrinterParser`](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L3312) used when calling `appendOffsetId()` [sets the offset in a separate field](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L3408) (which we see it's not respected as pointed in the bug itself).

Reply:
BTW, the one thing that this parser does besides parsing offsets is to [also parse zone strings](https://github.com/adoptium/jdk8u/blob/31b88042fba46e87fba83e8bfd43ae0ecb5a9afd/jdk/src/share/classes/java/time/format/DateTimeFormatterBuilder.java#L1076-L1077) (i.e. `Canada/Toronto`). I believe that would not be harmful, but maybe a test using a textual zone string may suffice?

Reply:
A backport request from Java 9 to 8 to fix this is still open
https://bugs.openjdk.org/browse/JDK-8149540